File: mcp/src/tools/signals.ts
/**
 * tools/signals.ts
 * evor_signal_emit   — dedup-upsert a Signal onto the run's signal bus
 * evor_signal_query  — pull signals by facet/severity/kind/tick lens (drains inbox first)
 * evor_signal_digest — compact top-slice for spawn-prompt injection
 *
 * Disk layout: <run_dir>/signals.jsonl (one JSON object per line).
 * Dedup key: `signature`. Repeat emits aggregate: occurrences+1,
 * last_seen=now, confidence raised toward 1.0, severity = MAX(old,new),
 * shapes/axes union, evidence merged. Mirrors harness/evor/signals.py exactly.
 *
 * Inbox drain (fix §15B parity): querySignals drains signals-inbox.jsonl
 * before loading so hook-captured signals written since the last query are
 * visible immediately — matches Python SignalBus.query() drain_inbox() call.
 * Pure TS atomic-rename drain — no Python subprocess dependency.
 */

import { createHash } from "crypto";
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import {
  Signal,
  SignalAxisSchema,
  SignalSchema,
  SignalShapeSchema,
  SignalSeveritySchema,
} from "../contracts.js";
import { resolveRunPaths } from "../run-store.js";
import { withRunLock } from "../lock.js";

// ── Constants ────────────────────────────────────────────────────────────────

const SEVERITY_ORDER: Record<string, number> = {
  low: 0,
  medium: 1,
  high: 2,
  critical: 3,
};

// ── Helpers ──────────────────────────────────────────────────────────────────

function nowIso(): string {
  return new Date().toISOString();
}

function signalId(kind: string, signature: string): string {
  const h = createHash("sha256").update(`${kind}:${signature}`).digest("hex").slice(0, 12);
  return `sig-${h}`;
}

function loadSignals(signalsPath: string): Signal[] {
  if (!existsSync(signalsPath)) return [];
  const out: Signal[] = [];
  for (const line of readFileSync(signalsPath, "utf8").split("\n")) {
    const t = line.trim();
    if (!t) continue;
    try {
      out.push(SignalSchema.parse(JSON.parse(t)));
    } catch {
      /* skip malformed lines */
    }
  }
  return out;
}

function atomicWriteJsonl(signalsPath: string, records: Signal[]): void {
  const dir = dirname(signalsPath);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  const tmpPath = `${signalsPath}.tmp`;
  writeFileSync(tmpPath, records.map((r) => JSON.stringify(r)).join("\n") + "\n", "utf8");
  renameSync(tmpPath, signalsPath);
}

/**
 * Drain <run_dir>/signals-inbox.jsonl into signals.jsonl.
 *
 * Mirrors Python SignalBus.drain_inbox(): atomically claims the inbox via
 * rename (crash-safe — a pre-existing .drain-tmp orphan is cleaned first),
 * then emits each line through emitSignal() (which dedups by signature).
 * Malformed lines are skipped silently.
 *
 * This is a pure-TS implementation so querySignals remains Python-free while
 * providing parity with the Python SignalBus.query() that calls drain_inbox()
 * before loading the bus.
 */
function drainSignalsInbox(runDir: string, runId: string, missionId?: string): void {
  const inboxPath = join(runDir, "signals-inbox.jsonl");
  if (!existsSync(inboxPath)) return;

  const tmpPath = `${inboxPath}.drain-tmp`;

  // Clean any orphaned .drain-tmp left by a prior crash
  if (existsSync(tmpPath)) {
    try { unlinkSync(tmpPath); } catch { /* ignore */ }
  }

  // Atomically claim the inbox before reading so a concurrent drain doesn't
  // process the same lines twice.
  try {
    renameSync(inboxPath, tmpPath);
  } catch {
    return; // inbox vanished (race) or rename failed — skip drain
  }

  let lines: string[];
  try {
    lines = readFileSync(tmpPath, "utf8").split("\n");
  } catch {
    return; // couldn't read drain-tmp — skip
  } finally {
    try { unlinkSync(tmpPath); } catch { /* ignore */ }
  }

  for (const line of lines) {
    const t = line.trim();
    if (!t) continue;
    try {
      const entry = JSON.parse(t) as Record<string, unknown>;
      emitSignal(
        runId,
        {
          kind: String(entry.kind ?? ""),
          signature: String(entry.signature ?? ""),
          shapes: Array.isArray(entry.shapes) ? (entry.shapes as string[]) : [],
          axes: Array.isArray(entry.axes) ? (entry.axes as string[]) : [],
          severity: String(entry.severity ?? "medium"),
          evidence: (entry.evidence as Record<string, unknown>) ?? {},
          source: String(entry.source ?? "hook:unknown"),
          tick: typeof entry.tick === "number" ? entry.tick : null,
          node_id: typeof entry.node_id === "string" ? entry.node_id : null,
          confidence: typeof entry.confidence === "number" ? entry.confidence : 0.5,
        },
        missionId,
      );
    } catch {
      /* skip malformed inbox lines */
    }
  }
}

// ── Core logic (exported for tests) ─────────────────────────────────────────

export interface EmitInput {
  kind: string;
  signature: string;
  shapes: string[];
  axes: string[];
  severity: string;
  evidence: Record<string, unknown>;
  source: string;
  tick?: number | null;
  node_id?: string | null;
  /** Initial confidence for new signals (0.0–1.0). Default 0.5. Mirrors make_signal() in signals.py. */
  confidence?: number;
}

/**
 * Persist or aggregate a signal on the run's bus. Dedup key = signature.
 *
 * Repeat emits: occurrences+1, last_seen=now, confidence raised toward 1.0
 * by (1-confidence)*0.4, severity = MAX(old, new), shapes/axes union,
 * evidence merged (incoming keys win on collision).
 *
 * `atomicWriteJsonl` is atomic per-file (tmp+rename), but the read-modify-write
 * is not locked, so a concurrent writer can clobber a just-written signal.
 * We defend against that by verifying the signal persisted after each write
 * and retrying — mirrors the upsertNode pattern in tree-store.ts.
 */
export function emitSignal(
  runId: string,
  input: EmitInput,
  missionId?: string,
): Signal {
  const paths = resolveRunPaths(runId, missionId);
  return withRunLock(paths.runDir, (): Signal => {
    const MAX_ATTEMPTS = 3;

    for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
      const existing = loadSignals(paths.signalsPath);

      const idx = existing.findIndex((s) => s.signature === input.signature);
      let final: Signal;

      if (idx !== -1) {
        const old = existing[idx];
        const newConf = Math.min(1.0, old.confidence + (1.0 - old.confidence) * 0.4);
        const newSev =
          SEVERITY_ORDER[input.severity] > SEVERITY_ORDER[old.severity]
            ? (input.severity as Signal["severity"])
            : old.severity;

        // Union shapes and axes, sort for stable output (mirrors Python sorted())
        const mergedShapes = [...new Set([...old.shapes, ...input.shapes])].sort() as Signal["shapes"];
        const mergedAxes = [...new Set([...old.axes, ...input.axes])].sort() as Signal["axes"];

        final = {
          signal_id: old.signal_id,
          kind: old.kind,
          signature: old.signature,
          shapes: mergedShapes,
          axes: mergedAxes,
          severity: newSev,
          evidence: { ...old.evidence, ...input.evidence },
          source: input.source,
          tick: input.tick !== undefined && input.tick !== null ? input.tick : old.tick,
          node_id: input.node_id !== undefined && input.node_id !== null ? input.node_id : old.node_id,
          confidence: Math.round(newConf * 10000) / 10000,
          occurrences: old.occurrences + 1,
          first_seen: old.first_seen,
          last_seen: nowIso(),
        };
        existing[idx] = final;
      } else {
        const now = nowIso();
        final = {
          signal_id: signalId(input.kind, input.signature),
          kind: input.kind,
          signature: input.signature,
          shapes: [...input.shapes].sort() as Signal["shapes"],
          axes: [...input.axes].sort() as Signal["axes"],
          severity: input.severity as Signal["severity"],
          evidence: input.evidence,
          source: input.source,
          tick: input.tick ?? undefined,
          node_id: input.node_id ?? undefined,
          confidence: input.confidence ?? 0.5,
          occurrences: 1,
          first_seen: now,
          last_seen: now,
        };
        existing.push(final);
      }

      atomicWriteJsonl(paths.signalsPath, existing);

      // Verify the signal actually persisted (guard against concurrent clobber).
      try {
        if (loadSignals(paths.signalsPath).some((s) => s.signature === final.signature)) {
          return final;
        }
      } catch {
        // file momentarily unreadable, retry
      }
    }

    throw new Error(
      `emitSignal: signal "${input.signature}" failed to persist in signals.jsonl after 3 attempts (concurrent clobber detected)`,
    );
  });
}

export interface QueryParams {
  shapes?: string[];
  axes?: string[];
  kind?: string;
  min_severity?: string;
  since_tick?: number;
  /** Cap the number of results returned (default 100). Guards agents against pulling the full bus. */
  limit?: number;
}

/**
 * Return signals matching a lens's subscription.
 *
 * Drains signals-inbox.jsonl first (parity fix §15B) so any hook captures
 * written since the last query are visible immediately — matches Python
 * SignalBus.query() which calls drain_inbox() before loading.
 *
 * Facet match is ANY-overlap: a signal matches if it shares >=1 requested
 * shape (when shapes given) AND >=1 requested axis (when axes given).
 * Sorted by (severity, confidence, last_seen) descending — highest-priority
 * first so digests take the top slice.
 */
export function querySignals(
  runId: string,
  params: QueryParams,
  missionId?: string,
): Signal[] {
  const paths = resolveRunPaths(runId, missionId);

  // Drain inbox before loading — parity with Python SignalBus.query()
  drainSignalsInbox(paths.runDir, runId, missionId);

  const floor = SEVERITY_ORDER[params.min_severity ?? "low"] ?? 0;

  const out = loadSignals(paths.signalsPath).filter((s) => {
    if (SEVERITY_ORDER[s.severity] < floor) return false;
    if (params.kind !== undefined && s.kind !== params.kind) return false;
    if (
      params.since_tick !== undefined &&
      (s.tick === null || s.tick === undefined || s.tick < params.since_tick)
    )
      return false;
    if (params.shapes && params.shapes.length > 0) {
      if (!params.shapes.some((sh) => (s.shapes as string[]).includes(sh))) return false;
    }
    if (params.axes && params.axes.length > 0) {
      if (!params.axes.some((ax) => (s.axes as string[]).includes(ax))) return false;
    }
    return true;
  });

  out.sort((a, b) => {
    const sevDiff = SEVERITY_ORDER[b.severity] - SEVERITY_ORDER[a.severity];
    if (sevDiff !== 0) return sevDiff;
    if (b.confidence !== a.confidence) return b.confidence - a.confidence;
    return b.last_seen.localeCompare(a.last_seen);
  });

  // P2-13: cap to prevent agents from pulling the full bus in one call
  const cap = params.limit ?? 100;
  return out.slice(0, cap);
}

export interface DigestParams {
  shapes?: string[];
  axes?: string[];
  min_severity?: string;
  max_items?: number;
}

/**
 * Return a compact top-slice of the signal bus for spawn-prompt injection.
 *
 * Mirrors Python SignalBus.digest(): defaults to severity>=medium so
 * low-noise signals never flood a spawn digest. Drains inbox via querySignals.
 */
export function digestSignals(
  runId: string,
  params: DigestParams,
  missionId?: string,
): Array<Record<string, unknown>> {
  const rows = querySignals(
    runId,
    {
      shapes: params.shapes,
      axes: params.axes,
      min_severity: params.min_severity ?? "medium",
    },
    missionId,
  ).slice(0, params.max_items ?? 8);

  return rows.map((s) => ({
    kind: s.kind,
    shapes: s.shapes,
    axes: s.axes,
    severity: s.severity,
    occurrences: s.occurrences,
    evidence: s.evidence,
  }));
}

// ── Tool registrations ───────────────────────────────────────────────────────

export function registerSignalTools(server: McpServer): void {
  // ── evor_signal_emit ────────────────────────────────────────────────────────
  server.tool(
    "evor_signal_emit",
    "Emit (upsert/dedup by signature) a Signal onto the run's signal bus at <run_dir>/signals.jsonl. " +
    "Repeat emits with the same signature aggregate: occurrences+1, confidence raised toward 1.0, severity escalates to MAX seen.",
    {
      run_id: z.string().describe("Active run identifier"),
      mission_id: z.string().optional().describe("Mission identifier (resolved automatically when omitted)"),
      kind: z.string().describe("Free-text signal type (e.g. 'cuda-oom', 'training-too-slow')"),
      signature: z.string().describe("Dedup key — identical signatures aggregate"),
      shapes: z
        .array(SignalShapeSchema)
        .min(1)
        .describe("Closed facet set: limit | opportunity | failure | trend"),
      axes: z
        .array(SignalAxisSchema)
        .describe("Closed facet set: memory | compute | accuracy | stability | data | generalization | cost"),
      severity: SignalSeveritySchema.describe("low | medium | high | critical"),
      evidence: z.record(z.string(), z.unknown()).describe("Structured evidence dict"),
      source: z.string().describe("Emitting role (e.g. 'evor-forge-analyst')"),
      tick: z.number().int().optional().describe("Optional evolution tick"),
      node_id: z.string().optional().describe("Optional associated node id"),
      confidence: z
        .number()
        .min(0)
        .max(1)
        .optional()
        .describe("Initial confidence for new signals (0.0–1.0). Default 0.5. Ignored on dedup-aggregate."),
    },
    async ({ run_id, mission_id, kind, signature, shapes, axes, severity, evidence, source, tick, node_id, confidence }) => {
      const missionId = mission_id ?? process.env.EVOR_MISSION_ID;
      const signal = emitSignal(
        run_id,
        { kind, signature, shapes, axes, severity, evidence, source, tick, node_id, confidence },
        missionId,
      );
      return {
        content: [{ type: "text" as const, text: JSON.stringify(signal) }],
      };
    },
  );

  // ── evor_signal_query ───────────────────────────────────────────────────────
  server.tool(
    "evor_signal_query",
    "Pull signals from the run's signal bus filtered by facet lens (ANY-overlap), severity floor, optional kind, and optional since_tick. " +
    "Drains signals-inbox.jsonl first so hook-captured signals are always visible. " +
    "Returns list sorted by (severity, confidence, last_seen) descending.",
    {
      run_id: z.string().describe("Active run identifier"),
      mission_id: z.string().optional().describe("Mission identifier (resolved automatically when omitted)"),
      shapes: z
        .array(SignalShapeSchema)
        .optional()
        .describe("ANY-overlap shape filter; omit to match all shapes"),
      axes: z
        .array(SignalAxisSchema)
        .optional()
        .describe("ANY-overlap axis filter; omit to match all axes"),
      kind: z.string().optional().describe("Exact kind filter; omit for all kinds"),
      min_severity: SignalSeveritySchema.default("low").describe("Severity floor (inclusive). Default: low"),
      since_tick: z.number().int().optional().describe("Only include signals with tick >= this value"),
      limit: z.number().int().min(1).max(500).optional().describe(
        "Cap on returned signals sorted by (severity, confidence, recency) desc. Default 100. "
        + "Use a smaller value (e.g. 10) for spawn-prompt injection to avoid context bloat.",
      ),
    },
    async ({ run_id, mission_id, shapes, axes, kind, min_severity, since_tick, limit }) => {
      const missionId = mission_id ?? process.env.EVOR_MISSION_ID;
      const signals = querySignals(
        run_id,
        { shapes, axes, kind, min_severity, since_tick, limit },
        missionId,
      );
      return {
        content: [{ type: "text" as const, text: JSON.stringify(signals) }],
      };
    },
  );

  // ── evor_signal_digest ──────────────────────────────────────────────────────
  server.tool(
    "evor_signal_digest",
    "Return a compact top-slice of the signal bus for spawn-prompt injection. " +
    "Defaults to severity>=medium so low-noise signals never flood a digest. " +
    "Equivalent to Python SignalBus.digest(). Drains inbox first.",
    {
      run_id: z.string().describe("Active run identifier"),
      mission_id: z.string().optional().describe("Mission identifier (resolved automatically when omitted)"),
      shapes: z
        .array(SignalShapeSchema)
        .optional()
        .describe("ANY-overlap shape filter; omit to match all shapes"),
      axes: z
        .array(SignalAxisSchema)
        .optional()
        .describe("ANY-overlap axis filter; omit to match all axes"),
      min_severity: SignalSeveritySchema.default("medium").describe(
        "Severity floor (inclusive). Default: medium (low kept out of digests by default)",
      ),
      max_items: z
        .number()
        .int()
        .min(1)
        .max(50)
        .default(8)
        .describe("Maximum signals to return (default 8)"),
    },
    async ({ run_id, mission_id, shapes, axes, min_severity, max_items }) => {
      const missionId = mission_id ?? process.env.EVOR_MISSION_ID;
      const digest = digestSignals(run_id, { shapes, axes, min_severity, max_items }, missionId);
      return {
        content: [{ type: "text" as const, text: JSON.stringify({ run_id, digest }) }],
      };
    },
  );
}

File: harness/evor/signals.py
"""
SignalBus — the run's neutral observation/pain-point bus.

Layout on disk:
  <run_dir>/signals.jsonl        — append/dedup store of Signal records

Design (see agents' <Signal_Lens> + the signal-protocol reference):
  - Producers EMIT neutral, self-describing signals (open `kind`, structured
    `evidence`, closed `shapes`/`axes` facets).
  - Consumers PULL by facet through their own lens; the same signal is a brief
    to Mutagen, a gate to Selector, a default to Forge-architect, an escalate to
    Evor — the bus never decides routing.

Storm / oscillation dampers:
  - dedup by `signature`: repeat emits increment occurrences, bump last_seen,
    raise confidence toward 1.0 (recurrence x confidence weighting).
  - severity gates whether a signal reaches a spawn digest.
  - query()/digest() sort by (severity, confidence, recency) so a one-off low
    signal never drowns a recurring high one.
"""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from .runlock import run_lock
from .state_root import assert_outside_plugin_root
from evor.contracts import CAPABILITY_GAP_KIND, Signal

_SEVERITY_ORDER = {"low": 0, "medium": 1, "high": 2, "critical": 3}
_INBOX_FILENAME = "signals-inbox.jsonl"


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _signal_id(kind: str, signature: str) -> str:
    h = hashlib.sha256(f"{kind}:{signature}".encode()).hexdigest()[:12]
    return f"sig-{h}"


def _atomic_write_jsonl(path: Path, lines: list[str]) -> None:
    """Atomically write the bus, PRESERVING records this writer never saw.

    Finding O-02, item 1.8. This is the commit point of a read-modify-write:
    ``emit`` loads every signal, merges one, and writes the whole list back. Any
    record another writer committed between that load and this write was silently
    erased — the file is rewritten wholesale, so a lost update leaves no trace at
    all, in a bus whose entire purpose is to not lose signals.

    The run lock excludes the other PROCESS that writes this file. It cannot
    exclude what reaches the file by another route: an appender, a re-entrant
    call, a writer that does not take the lock. So the union happens HERE, at the
    moment of commit, where the window is empty rather than merely narrow:
    re-read the file, keep every signature the caller is not writing, then
    replace. Records the caller IS writing win — it has just merged them.

    Bus-only by construction (``gotchas.py`` has its own writer). A generic
    atomic writer would not merge; this one is named for its file.
    """
    path.parent.mkdir(parents=True, exist_ok=True)

    incoming: dict[str, str] = {}
    order: list[str] = []
    for line in lines:
        try:
            sig = json.loads(line).get("signature")
        except (ValueError, AttributeError):
            sig = None
        key = sig if sig is not None else f"__unkeyed__{len(order)}"
        if key not in incoming:
            order.append(key)
        incoming[key] = line

    if path.exists():
        try:
            for raw in path.read_text().splitlines():
                raw = raw.strip()
                if not raw:
                    continue
                try:
                    sig = json.loads(raw).get("signature")
                except (ValueError, AttributeError):
                    continue
                if sig is not None and sig not in incoming:
                    incoming[sig] = raw
                    order.append(sig)
        except OSError:
            # Unreadable bus — writing what we have beats refusing to write.
            pass

    fd, tmp = tempfile.mkstemp(dir=path.parent, suffix=".tmp")
    try:
        with os.fdopen(fd, "w") as fh:
            for key in order:
                fh.write(incoming[key] + "\n")
        os.replace(tmp, path)
    except Exception:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


def _load(path: Path) -> list[Signal]:
    if not path.exists():
        return []
    out: list[Signal] = []
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            out.append(Signal.model_validate_json(line))
        except Exception:
            continue
    return out


class SignalBus:
    """Append/dedup/query store for Signal records on a run's bus."""

    def __init__(self, run_dir: Path) -> None:
        self._run_dir = Path(run_dir)
        self._path = self._run_dir / "signals.jsonl"

    # ── Emit ────────────────────────────────────────────────────────────────────
    def emit(self, signal: Signal) -> Signal:
        """Persist or aggregate a signal, under the shared run lock (item 1.8).

        The body is a read-modify-write: load every signal, merge this one by
        signature, rewrite the file. Two concurrent emits lose one — and there
        genuinely are two writers, in two languages: the Python drain called from
        `subagent-stop.mjs` and the MCP server's own signal tools, which already
        take `<run_dir>/.tree.lock` through `withRunLock`. Taking the same lock
        here is what makes that mutual.

        `signals.jsonl` is one of exactly two files this release locks. The
        ownership rule is in `runlock.py`: a lock exists only where a second
        writer genuinely survives in another process. Twelve pairwise retrofits
        were considered and cut — a lock around a single-writer file buys nothing
        and adds a stale-lock failure mode.
        """
        # P-02: refuse at the WRITE, which is the thing that must not happen.
        # Checking at construction was tried and is worse: a bus that is built
        # and never emits writes nothing, so refusing to build it denies an act
        # that was never going to occur — and it moves the error to a line that
        # is not the one doing anything wrong.
        assert_outside_plugin_root(self._run_dir, "the signal bus")
        with run_lock(self._run_dir):
            return self._emit_locked(signal)

    def _emit_locked(self, signal: Signal) -> Signal:
        """The critical section of :meth:`emit`. Never call without the lock."""
        existing = _load(self._path)
        idx: Optional[int] = None
        for i, s in enumerate(existing):
            if s.signature == signal.signature:
                idx = i
                break

        if idx is not None:
            old = existing[idx]
            new_conf = min(1.0, old.confidence + (1.0 - old.confidence) * 0.4)
            sev = old.severity
            if _SEVERITY_ORDER[signal.severity] > _SEVERITY_ORDER[sev]:
                sev = signal.severity
            merged = Signal(
                signal_id=old.signal_id,
                kind=old.kind,
                signature=old.signature,
                shapes=sorted(set(old.shapes) | set(signal.shapes)),
                axes=sorted(set(old.axes) | set(signal.axes)),
                severity=sev,
                evidence={**old.evidence, **signal.evidence},
                source=signal.source,
                tick=signal.tick if signal.tick is not None else old.tick,
                node_id=signal.node_id or old.node_id,
                confidence=round(new_conf, 4),
                occurrences=old.occurrences + 1,
                first_seen=old.first_seen,
                last_seen=_now_iso(),
            )
            existing[idx] = merged
            final = merged
        else:
            final = Signal(
                signal_id=_signal_id(signal.kind, signal.signature),
                kind=signal.kind,
                signature=signal.signature,
                shapes=signal.shapes,
                axes=signal.axes,
                severity=signal.severity,
                evidence=signal.evidence,
                source=signal.source,
                tick=signal.tick,
                node_id=signal.node_id,
                confidence=signal.confidence,
                occurrences=signal.occurrences,
                first_seen=signal.first_seen or _now_iso(),
                last_seen=signal.last_seen or _now_iso(),
            )
            existing.append(final)

        # The commit point unions with whatever is on disk (see
        # `_atomic_write_jsonl`), so a record another writer added inside this
        # read-modify-write window survives. One mechanism, at the moment where
        # the window is empty rather than merely narrow.
        _atomic_write_jsonl(self._path, [s.model_dump_json() for s in existing])
        return final

    # ── Query (the pull half) ────────────────────────────────────────────────────
    def query(
        self,
        shapes: Optional[list[str]] = None,
        axes: Optional[list[str]] = None,
        kind: Optional[str] = None,
        min_severity: str = "low",
        since_tick: Optional[int] = None,
    ) -> list[Signal]:
        """Return signals matching a lens's subscription.

        Lazily drains signals-inbox.jsonl before loading, so any hook captures
        written since the last query are visible immediately.

        Facet match is ANY-overlap: a signal matches if it shares >=1 requested
        shape (when shapes given) AND >=1 requested axis (when axes given).
        Sorted by (severity, confidence, last_seen) descending — highest-priority
        first, so digests take the top slice.
        """
        drain_inbox(self._run_dir, self)
        floor = _SEVERITY_ORDER.get(min_severity, 0)
        out: list[Signal] = []
        for s in _load(self._path):
            if _SEVERITY_ORDER[s.severity] < floor:
                continue
            if kind is not None and s.kind != kind:
                continue
            if since_tick is not None and (s.tick is None or s.tick < since_tick):
                continue
            if shapes and not (set(s.shapes) & set(shapes)):
                continue
            if axes and not (set(s.axes) & set(axes)):
                continue
            out.append(s)
        out.sort(
            key=lambda s: (_SEVERITY_ORDER[s.severity], s.confidence, s.last_seen),
            reverse=True,
        )
        return out

    def digest(
        self,
        shapes: Optional[list[str]] = None,
        axes: Optional[list[str]] = None,
        min_severity: str = "medium",
        max_items: int = 8,
    ) -> list[dict[str, Any]]:
        """Compact top-slice for a spawn prompt (the mandatory PUSH half).

        Defaults to severity>=medium so low-noise never floods a digest.
        """
        rows = self.query(shapes=shapes, axes=axes, min_severity=min_severity)[:max_items]
        return [
            {
                "kind": s.kind,
                "shapes": s.shapes,
                "axes": s.axes,
                "severity": s.severity,
                "occurrences": s.occurrences,
                "evidence": s.evidence,
            }
            for s in rows
        ]


def make_signal(
    kind: str,
    signature: str,
    shapes: list[str],
    axes: list[str],
    severity: str,
    evidence: dict[str, Any],
    source: str,
    tick: Optional[int] = None,
    node_id: Optional[str] = None,
    confidence: float = 0.5,
) -> Signal:
    """Convenience factory for a Signal with sensible defaults."""
    now = _now_iso()
    return Signal(
        signal_id=_signal_id(kind, signature),
        kind=kind,
        signature=signature,
        shapes=shapes,  # type: ignore[arg-type]
        axes=axes,  # type: ignore[arg-type]
        severity=severity,  # type: ignore[arg-type]
        evidence=evidence,
        source=source,
        tick=tick,
        node_id=node_id,
        confidence=confidence,
        occurrences=1,
        first_seen=now,
        last_seen=now,
    )


def drain_inbox(run_dir: Path, bus: "SignalBus | None" = None) -> int:
    """Drain <run_dir>/signals-inbox.jsonl into the SignalBus.

    Reads each line of the inbox file, builds a Signal via make_signal(),
    calls bus.emit() (which dedups by signature and updates occurrences/
    confidence/last_seen), then removes the inbox file.  Idempotent and
    crash-safe: the inbox is atomically renamed before processing so a crash
    mid-drain leaves a *.drain-tmp orphan (not the live inbox), meaning the
    next drain sees an empty inbox and produces no duplicates.  Malformed
    lines are skipped silently — one bad entry never blocks the rest.

    Inbox line schema (produced by post-tool-use.mjs):
      {kind, signature, shapes, axes, severity, evidence, source, created_at}

    Returns the count of successfully emitted signals.
    """
    inbox = Path(run_dir) / _INBOX_FILENAME
    if not inbox.exists():
        return 0
    if bus is None:
        bus = SignalBus(run_dir)

    # Atomically claim the inbox before reading so a concurrent or re-entrant
    # drain doesn't process the same lines twice.
    fd, tmp = tempfile.mkstemp(dir=inbox.parent, suffix=".drain-tmp")
    os.close(fd)
    try:
        os.replace(str(inbox), tmp)
    except FileNotFoundError:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        return 0

    # ── O-17: an interrupted drain must not eat the batch it claimed ────────
    #
    # The claimed batch used to be unlinked in a `finally`, so anything the loop
    # had not reached was destroyed — and the interruption that matters here is
    # a `KeyboardInterrupt` or an OOM kill, which are BaseExceptions and sail
    # straight past `except Exception: continue`. Claiming the inbox by rename
    # is right; deleting the claim before the work is done is not.
    #
    # The remainder is written BACK to the inbox, so a later drain delivers it.
    # Re-delivering an already-emitted signal is harmless — the bus dedups by
    # signature and counts occurrences — while losing one is not, which is what
    # makes returning the whole remainder the safe direction.
    lines = [ln.strip() for ln in Path(tmp).read_text().splitlines() if ln.strip()]
    emitted = 0
    try:
        for index, raw in enumerate(lines):
            try:
                entry = json.loads(raw)
                sig = make_signal(
                    kind=entry["kind"],
                    signature=entry["signature"],
                    shapes=list(entry.get("shapes") or []),
                    axes=list(entry.get("axes") or []),
                    severity=entry.get("severity", "medium"),
                    evidence=entry.get("evidence") or {},
                    source=entry.get("source", "hook:unknown"),
                )
                bus.emit(sig)
                emitted += 1
            except Exception:
                # A malformed line is skipped, as before: one bad entry must
                # never block the rest of the batch.
                continue
    except BaseException:
        # Interrupted. Return everything from the line that failed onward.
        _restore_inbox(inbox, lines[index:])
        raise
    finally:
        try:
            os.unlink(tmp)
        except OSError:
            pass

    return emitted


def _restore_inbox(inbox: Path, remaining: list[str]) -> None:
    """Put un-drained lines back, ahead of anything written since (O-17)."""
    if not remaining:
        return
    try:
        existing = inbox.read_text().splitlines() if inbox.exists() else []
    except OSError:
        existing = []
    try:
        inbox.write_text("\n".join([*remaining, *[ln for ln in existing if ln.strip()]]) + "\n")
    except OSError:
        # Nothing further can be done; the signals are lost either way, but the
        # exception that caused this is about to propagate and say so.
        pass


def open_capability_gaps(run_dir: Path) -> list[Signal]:
    """Capability gaps currently blocking this run (item 2b.1).

    A `capability-gap` says an agent was asked to do something it cannot do —
    `evor-tick` emitted `forge-cannot-spawn-forge-junior-tool-gap`, correctly and
    at the right moment, and nothing read it, so the tick spun until a human
    noticed and restarted the mission.

    Returned newest first, because the most recent gap is the one blocking now.
    """
    try:
        bus = SignalBus(run_dir)
    except Exception:  # noqa: BLE001 — a bus that refuses to load blocks nothing
        return []
    gaps = [s for s in bus.query() if s.kind == CAPABILITY_GAP_KIND]
    gaps.sort(key=lambda s: s.last_seen or "", reverse=True)
    return gaps
